### DISCUSSION TITLE
How to debug redis

### DISCUSSION THREAD

**Discussion by ngjaying** [lf-edge/ekuiper#978] [General]
@rwadowski In my smoke testing with the redis store for qos 2, I got some errors like below randomly:
```
time="2021-08-26 17:45:11" level=info msg="Cannot save checkpoint 1629971111464 due to storage error: save checkpoint err: list at KV:TS:rule1 key should be non empty" file="checkpoint/coordinator.go:257" rule=rule1
```
I believe the cause is in `internal/pkg/ts/redis/redisTs.go` Set function:
```go
reply, err := conn.Do(AddToSortedSet, t.key, key, b)
if err != nil {
	return err
}
length, err := redis.Int(reply, err)
if err != nil {
	return err
}
if length == 0 {
	return fmt.Errorf("list at %s key should be non empty", t.key)
}
t.last = key
return nil
```

The reply is 0 sometimes. How can we know why it has no error but return a 0?

**Comment by rwadowski** [comment]
@ngjaying There is nothing about case when given score already exists in the sorted set in the [documentation](https://redis.io/commands/ZADD). Is it possible to save checkpoint with same timestamp ?

I made small test
```
127.0.0.1:6379> ZADD test-set 1 "one"
(integer) 1
127.0.0.1:6379> ZADD test-set 1 "one"
(integer) 0
127.0.0.1:6379> ZADD test-set 1 "one1"
(integer) 1
127.0.0.1:6379> ZADD test-set 1 "one1"
(integer) 0
127.0.0.1:6379> ZRANGE test-test 0 -1
(empty list or set)
127.0.0.1:6379> ZRANGE test-set 0 -1
1) "one"
2) "one1"
```
It looks like that zero value is returned in the case of the same score

**Reply by ngjaying** [reply]
No, there should be no duplicate. How to check if that score(timestamp) exist in redis

**Reply by rwadowski** [reply]
There are flags NX (add only new elements) or GT (only update existing elements if the new score is greater than current one) -> [ZADD options](https://redis.io/commands/ZADD#zadd-options)
I played a little bit with options. Although it returns 0 it is fine since it updates the score for the given member.By default 0 is returned because element is not added, it is updated (its score). I think code should not check return value since this scenario is possible.

```
127.0.0.1:6379> ZADD k0 1 "first"
(integer) 1
127.0.0.1:6379> ZADD k0 2 "second"
(integer) 1
127.0.0.1:6379> ZADD k0 3 "third"
(integer) 1
127.0.0.1:6379> ZRANGE k0 0 -1 WITHSCORES
1) "first"
2) "1"
3) "second"
4) "2"
5) "third"
6) "3"
127.0.0.1:6379> ZADD k0 4 "third"
(integer) 0
127.0.0.1:6379> ZRANGE k0 0 -1 WITHSCORES
1) "first"
2) "1"
3) "second"
4) "2"
5) "third"
6) "4"
```

**Reply by rwadowski** [reply]
Does logic require return an error in case of duplicate ?

**Reply by rwadowski** [reply]
> No, there should be no duplicate. How to check if that score(timestamp) exist in redis

func (t ts) Last(value interface{}) (int64, error) which uses ZREVRANGE key 0 0 WITHSCORES

**Reply by ngjaying** [reply]
"Does logic require return an error in case of duplicate ?"  No. And it seems there should be no duplicate. Especially in the set method, you even check the last timestamp.
```go
if key <= t.last {
		return false, nil
	}
```

**Reply by rwadowski** [reply]
Is it possible to call this interface from many threads ?

**Reply by rwadowski** [reply]
So maybe we should just allow to overwrite ? It is nosql after all
